### DISCUSSION TITLE
Custom Prompt for Active Note index?

### DISCUSSION THREAD

**Discussion by jetsquared** [logancyang/obsidian-copilot#74] [Ideas]
Thank you for this amazing plugin!

My current workflow is as such:
- open an active document
- {click} Rebuild Index for Active Document
- Change Mode Selection to QA: Active Note
- Enter my custom prompt and press send.

Q1 - once the user has {clicked} selected to index the Active Document, can the "QA: Active Note" be switched to automatically?
Q2 - Can a "Custom User Prompt" be stored for a "QA: Active Document"?  For example: in an active document, a user would initiate the custom prompt, copilot would begin indexing the active document, set the correct mode, insert the prompt, and run it.

**Comment by logancyang** [comment]
Hi @jetsquared , these are great suggestions! Thank you for your ideas, it helps me understand what actual usage looks like besides my own.

This is precisely why I put BETA next to QA mode, to learn what makes sense. I created the indexing button initially just for wrapping the active note in a single prompt to the chat model, so one click was enough to start the QA. Then I integrated with LangchainJS, and introduced retrieval QA chain for "QA: active note" mode. So basically, the two modes are running two different chains, hence the one-click QA with the button doesn't cut it anymore. The dropdown for mode selection was introduced.

I think these 2 points both make sense. The indexing button means the user wants to start QA, so there's no need to remain in conversation mode - automatically switching to QA should be intuitive. Once the user wants to stop the QA, just manually switch to conversation mode again.

The custom prompt for QA mode is quite interesting. Do you mean that this particular custom prompt is only meant for QA mode, as some prompts might not make sense in Conversation mode? I can introduce a checkbox for the prompt creation modal: `[ ] QA Mode only`, and implement the auto-indexing and switching mode at the time of applying the prompt.

Pls let me know if I understand correctly.

In the next few iterations, I plan to enhance QA mode to not just Active Note, but optionally a folder of notes where the path is set in the settings. Hopefully these proposed UX still make sense then! Any more ideas and suggestions are welcome!

**Reply by jetsquared** [reply]
First off your LangchainJS implementation is amazing.  Great work.  I've cloned and played with PrivateGPT and others, and they proved to be too intensive.  You have gold here.  I'm super excited about your project.

> ... user wants to start QA, so there's no need to remain in conversation mode - automatically switching to QA should be intuitive. Once the user wants to stop the QA, just manually switch to conversation mode again.

Yes, spot on.

> ... some prompts might not make sense in Conversation mode?

Now that you mention that... that made me rethink it.  The custom prompt could be used for smaller text selections too. 
So, I realize this is my own use case. So I am extremely grateful for your dialogue.

- I have a set of docs that I know I will use the same prompt on.  In reality, I'm fine tuning the prompt as I go.  Once I'm happy with it, I  use the same prompt on each of the docs.
- So... perhaps it is as simple as, `Copilot: Run custom prompt for active document`
  - `if (chatExists) { clearChatPrompt(); } //  clear current chat (ok/cancel)`
  - index doc, start QA, insert and run the prompt

Thank you for your response.  Thank you.
ALSO.  I've forked the repo and I'd love to contribute. I only worked with Angular at work (mainly .net for me), but I believe I can come up with some clean React 🤓. Any strong opinions for contributors? Let me know if you have any resources for best ways to contribute to your project through PRs, if that's ok?

**Reply by logancyang** [reply]
I myself am not a frontend person, still learning React by building this project. Contribution is welcome! In terms of Obsidian plugin dev, you can check out their docs: https://marcus.se.net/obsidian-plugin-docs/ and join their discord plugin-dev channel. I'm not active there though.

Just need to set up your local workflow to be able to test your changes effectively. This is a pretty good guide: https://medium.com/gitconnected/how-to-set-up-the-ideal-obsidian-plugin-development-workflow-b222fe72280f
